pkg/cri/resource-manager/policy/builtin/topology-aware/resources.go

```diff
@@ -830,6 +839,73 @@ func (cs *supply) DumpAllocatable() string {
 	return allocatable
 }
 
+// prettyMem formats the given amount as k, M, G, or T units.
+func prettyMem(amount uint64) string {
+	units := []string{"k", "M", "G", "T"}
+	a := float64(amount)
+	f := a
+	u := ""
+	for _, unit := range units {
+		if f /= float64(1024); f < 1 {
+			break
+		}
+		a = f
+		u = unit
+	}
+	return strconv.FormatFloat(a, 'f', 2, 64) + u
+}
+
+// DumpMemoryState dumps the state of the available and allocated memory.
+func (cs *supply) DumpMemoryState(prefix string) {
+	memTypes := []memoryType{memoryDRAM, memoryPMEM, memoryHBM}
+	totalFree := uint64(0)
+	totalGranted := uint64(0)
+	for _, kind := range memTypes {
+		free := cs.mem[kind]
+		granted := cs.grantedMem[kind]
+		if free != 0 || granted != 0 {
+			log.Debug(prefix+"- %s: free: %s, granted %s",
+				kind, prettyMem(free), prettyMem(granted))
+		}
```

Comment:
`free == 0 || granted != 0` would be impossible or a bug(?) Should we print something in that case?

Reply:
I'm not sure what you mean... Did you mean `free == 0 && granted != 0` ?

Anyway, the invariant wrt. free `(cs.mem[kind])` and granted (`cs.grantedMem[kind]`) for any node is that `free + granted == total available`. Whenever an allocation is done, that's worth is moved from free to granted (this was exactly one of the bugs, the inconsistent use of free) and the other way around when it is released.

Now with that in mind, pretty much any combinations of free and granted are possible (maybe with the exception of `free == 0 && granted == 0` since we don't have memory-less nodes in the tree).

Reply:
Argh, sorry, again I was reading too quickly. Or, maybe I should get a pair of glasses 😬   I was reading `&&` instead of `||`, sorry about that...